File: python/darvishi_barati.py
# M4 – Darvishi and Barati (2007), Articulo1.pdf pag 3

from sympy import Derivative, sympify
from python.graficar_error import graficar_error

"""
Metodo de Darvishi-Barati para encontrar el cero de una funcion
:param str_funcion: string con la funcion que se debe evaluar
:param xk: valor de x inicial con el cual aplicar el metodo
:param tol: tolerancia al fallo de debe tener el resultado final
:returns: lista con dos elementos, xk calculado y numero iteraciones
"""


def darvishi_barati(str_funcion, xk, tol, graph=1):
    # Se verifica el tipo de str_funcion
    if type(str_funcion) != str:
        return "La funcion debe estar en un string"

    # Se verifica el tipo de xk
    if type(xk) != int and type(xk) != float:
        return "xk debe ser un numero"

    # Se verifica el tipo de la tolerancia
    if type(tol) != int and type(tol) != float:
        return "La tolerancia debe ser un numero"

    # Se verifica que el valor de graph sea uno o cero
    if graph != 1 and graph != 0:
        return "graph debe ser cero (desactivado) o uno (activado)"

    try:
        funcion = sympify(str_funcion)  # Se obtiene la funcion del string
        df = Derivative(funcion, 'x')  # Se deriva la funcion
        itr = 0  # Se inicializa el contador de iteraciones

        # Listas donde se guardan los valores para graficar el error
        lista_fxk = []
        lista_iter = []

        while 1:
            fxk = float(funcion.subs({'x': xk}))  # Se evalua la funcion

            if graph == 1:  # Se guardan los valores para la grafica
                lista_fxk += [abs(fxk)]
                lista_iter += [itr]

            if abs(fxk) <= tol:  # Se verifica la condicion de parada
                break

            else:
                # Se calcula el valor de xk en la derivada de f
                df_xk = float(df.doit().subs({'x': xk}))

                if df_xk == 0:  # Se verifica para evitar division entre cero
                    break

                # Variable auxiliar para el calculo del numerador
                numerador_aux = float(xk - fxk / df_xk)

                # Se calcula el numerador
                numerador = float(funcion.subs({'x': numerador_aux}))

                xk = xk - (fxk / df_xk) - (numerador / df_xk)

                itr += 1

        if graph == 1:
            graficar_error(lista_iter, lista_fxk)

        return [xk, itr]

    except:
        return "Sintaxis de la funcion es incorrecta"

File: octave/darvishi_barati.m
pkg load symbolic

function [x_aprox, iter] = darvishi_barati(str_funcion, xk, tol, graph)
    % Metodo de Darvishi-Barati para encontrar el cero de una funcion
    % :param str_funcion: string con la funcion que se debe evaluar
    % :param xk: valor de x inicial con el cual aplicar el metodo
    % :param tol: tolerancia al fallo de debe tener el resultado final
    % :returns: xk calculado y numero iteraciones

    if nargin == 3  % Si el numero de argumento es igual a 3
        graph = 1   % Se declara con el valor por defecto
    end

    % Se verifica el tipo de dato de str_funcion
    if ~isa(str_funcion, 'char')
        disp("La funcion debe estar en un char")
        return  % Se finaliza la ejecucion
    end

    % Se verifica el tipo de dato de xk
    if ~isa(xk, 'double')
        disp("xk debe ser un numero")
        return  % Se finaliza la ejecucion
    end

    % Se verifica el tipo de dato de la tolerancia
    if ~isa(tol, 'double')
        disp("La tolerancia debe ser un numero")
        return  % Se finaliza la ejecucion
    end

    % Se verifica que el valor de graph sea uno o cero
    if graph ~= 1 && graph ~= 0
        disp("graph debe ser cero (desactivado) o uno (activado)")
        return  % Se finaliza la ejecucion
    end

    try
        funcion = matlabFunction(sym(str_funcion));   % Se obtiene la funcion
        df = matlabFunction(diff(sym(funcion)));  % Se calcula la derivada

        % Listas donde se guardaran los valores para la grafica de error
        lista_fxk = [];
        lista_iter = [];

        itr = 0;  % Se inicializa el contador de iteraciones

        while 1
            fxk = funcion(xk);  % Se evalua la funcion

            if graph == 1
                % Se guardan los valores para la grafica
                lista_fxk = [lista_fxk abs(fxk)]
                lista_iter = [lista_iter (itr)]
            end  % graph == 1

            if abs(fxk) <= tol  % Se verifica la condicion de parada
                break;

            else
                % Se calcula el valor de xk en la derivada de f
                df_xk = df(xk)

                if df_xk == 0  % Se verifica para evitar division entre cero
                    break
                end  % df_xk == 0

                % Calculo del numerador
                numerador = funcion(xk - fxk / df_xk)

                xk = xk - (fxk / df_xk) - (numerador / df_xk)
                itr = itr + 1

            end  % abs(f(xk)) <= tol

        end  % while 1

    catch
        disp("Sintaxis de la funcion es incorrecta")
        return  % Se finaliza la ejecucion

    end  % try / catch


    if graph == 1
        graficar_error(lista_iter, lista_fxk)
    end  % graph == 1

    x_aprox = xk
    iter = itr

end  % darvishi_barati(str_funcion, xk, tol, graph)